src/playback-watcher.js

```diff
@@ -126,6 +150,75 @@ export default class PlaybackWatcher {
       window.setTimeout(this.monitorCurrentTime_.bind(this), 250);
   }
 
+  /**
+   * Reset stalled download stats for a specific type of loader
+   *
+   * @param {string} type
+   *        Whether to check the audio or main segment loader.
+   */
+  resetSegmentDownloads_(type) {
+    const loader = this.masterPlaylistController_[`${type}SegmentLoader_`];
+
+    if (this[`${type}StalledDownloads_`] > 0) {
+      this.logger_(`reseting stalled downloads for ${type} loader`);
+    }
+    this[`${type}StalledDownloads_`] = 0;
+    this[`${type}Buffered_`] = loader.buffered_();
+  }
+
+  /**
+   * Checks on every main/audio segment `appended` to see
+   * if segment appends are making progress. If they are not
+   * and we are still downloading bytes. We blacklist the playlist.
+   *
+   * @param {string} type
+   *        Whether to check the audio or main segment loader.
+   */
+  checkSegmentDownloads_(type) {
+    const mpc = this.masterPlaylistController_;
+    const loader = mpc[`${type}SegmentLoader_`];
+    const buffered = loader.buffered_();
+    const isBufferedDifferent = Ranges.isRangeDifferent(this[`${type}Buffered_`], buffered);
+
+    this[`${type}Buffered_`] = buffered;
+
+    // if another watcher is going to fix the issue or
+    // the buffered value for this loader changed
+    // appends are working
+    if (isBufferedDifferent) {
+      this.resetSegmentDownloads_(type);
+      return;
+    }
+
+    this[`${type}StalledDownloads_`]++;
+
+    this.logger_(`found stalled download #${this[`${type}StalledDownloads_`]} for ${type} loader`);
+    // We will technically get past this on the fourth bad append
+    // rather than the third. As the first will almost always cause
+    // buffered to change which means that StalledDownloads_ will
+    // not be incremented
+    if (this[`${type}StalledDownloads_`] < 3) {
+      return;
+    }
+
+    this.logger_(`${type} loader download exclusion`);
+    this.resetSegmentDownloads_(type);
+    this.tech_.trigger({type: 'usage', name: `vhs-${type}-download-exclusion`});
+
+    if (type !== 'subtitle') {
+      mpc.blacklistCurrentPlaylist({
+        message: `Infinite ${type} segment downloading detected.`
+      }, Infinity);
+    } else {
```

Comment:
Subtitles cannot be excluded. So we have to disable and remove the track. If the track isn't appending to the buffer correctly anyway then its not going to be showing subtitles making the track useless.

Reply:
I agree that there could be accessibility implications and I will add a TODO.